File: Problem_1.py
#
# @lc app=leetcode id=75 lang=python3
#
# [75] Sort Colors
#

# @lc code=start
'''
Time Complexity - O(n). We are iterating the entire array only once
Space Complexity - O(1), We are using extra variables

This code works on leetcode
'''
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        low, mid, high =0, 0, len(nums)-1 #set low and mid to 0 and high to last index
        
        while mid <= high:
            if nums[mid]==2:
                nums[mid], nums[high] = nums[high], nums[mid] #swap elements at mid and high if we find a 2 at mid
                high-=1 #element at high is fixed move one index to left
            elif nums[mid] == 0: 
                nums[low], nums[mid] = nums[mid], nums[low] #swap elements at mid and low if we find a 0 at mid
                low+=1 #element at low is fixed moce one index to right
                mid+=1 #move mid to right as well else we may not move mid at all
            else:
                mid+=1
        
# @lc code=end

File: Problem_1.js
// Problem1 (https://leetcode.com/problems/sort-colors/)

// Time Complexity : O(n), n is the length of array
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
We use 3 ponters.
low points to 0
high points to 2
mid points to 1
*/

let swap = (nums, a, b) => {
    let temp = nums[a];
    nums[a] = nums[b];
    nums[b] = temp;
}
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
    if (nums === 0 || nums.length === 0) {
        return;
    }
    let low = 0;
    let high = nums.length - 1;
    let mid = 0;
    while (mid <= high) {
        // If mid is found, swap with high
        // Now high is 2 so high--
        if (nums[mid] === 2) {
            swap(nums, mid, high);
            high--;
        } else if (nums[mid] === 0) {
            // If mid === 0, swap with low now low has 0 so low++
            swap(nums, mid, low);
            mid++;
            low++;
        } else {
            mid++;
        }
    }
};